components/index_embeddings/src/main.py

```diff
@@ -0,0 +1,69 @@
+import logging
+
+import dask.dataframe as dd
+import weaviate
+from fondant.component import DaskWriteComponent
+from weaviate.util import _capitalize_first_letter
+
+logger = logging.getLogger(__name__)
+
+
+class IndexEmbeddingsComponent(DaskWriteComponent):
+    def __init__(
+        self,
+        *_,
+        weaviate_url: str,
+        model: str,
+        dataset: str,
+        vectorizer: dict,
+        overwrite: bool,
+    ):
+        self.client = weaviate.Client(weaviate_url)
+        self.overwrite = overwrite
+
+        # following weaviate convention for upper case class names
+        self.class_name = _capitalize_first_letter(f"{model}_{dataset}")
+        self.class_obj = {
+            "class": self.class_name,
+            "properties": [
+                {
+                    "name": "passage",
+                    "dataType": ["text"],
+                },
+                {  # id of the passage in the passage dataset not to mix up with weaviate's uuid
+                    "name": "identifier",
+                    "dataType": ["text"],
+                },
+            ],
+        }
+
+        self.class_obj.update(vectorizer)
+
+    def write(self, dataframe: dd.DataFrame):
+        # using weaviate instead of its langchain wrapper
+        # because langchain does not yet provide ingestion of pre-computed embeddings
+        if self.overwrite:
+            self.client.schema.delete_class(self.class_name)
+            self.client.schema.create_class(self.class_obj)
+        elif not self.client.schema.exists(self.class_name):
+            self.client.schema.create_class(self.class_obj)
+        else:
+            # do not index
+            return dataframe
+
+        self.client.batch.configure(batch_size=100, dynamic=True, num_workers=2)
+        with self.client.batch as batch:
+            for part in dataframe.partitions:
```

Comment:
instead of looping through each individual partition, can we use Dask's [`map_partitions()`](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.map_partitions.html) to write partitions in parallel. This would increase the speed quite significantly for large datasets. One caveat is to check if the weaviate DB can handle concurrent writes.

Reply:
This seems to introduce problems due to both `Dask` and `Weaviate` trying to spawn processes / threads.